sdk/core/Azure.Core/src/SequenceWriter.cs

```diff
@@ -0,0 +1,189 @@
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Azure.Core
+{
+    /// <summary>
+    /// A buffer writer which writes large sequences of data into smaller shared buffers.
+    /// </summary>
+    internal sealed class SequenceWriter : IBufferWriter<byte>, IDisposable
+    {
+        private struct Buffer
+        {
+            public byte[] Array;
+            public int Written;
+        }
+
+        private Buffer[] _buffers; // this is an array so items can be accessed by ref
+        private int _count;
+        private int _bufferSize;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="SequenceWriter"/>.
+        /// </summary>
+        /// <param name="segmentSize">The size of each buffer segment.</param>
+        public SequenceWriter(int segmentSize = 4096)
+        {
+            _bufferSize = segmentSize;
+            _buffers = Array.Empty<Buffer>();
+        }
+
+        /// <summary>
+        /// Notifies the <see cref="SequenceWriter"/> that bytes bytes were written to the output <see cref="Span{T}"/> or <see cref="Memory{T}"/>.
+        /// You must request a new buffer after calling <see cref="Advance(int)"/> to continue writing more data; you cannot write to a previously acquired buffer.
+        /// </summary>
+        /// <param name="bytesWritten">The number of bytes written to the <see cref="Span{T}"/> or <see cref="Memory{T}"/>.</param>
+        /// <exception cref="ArgumentOutOfRangeException"></exception>
+        public void Advance(int bytesWritten)
+        {
+            ref Buffer last = ref _buffers[_count - 1];
+            last.Written += bytesWritten;
+            if (last.Written > last.Array.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bytesWritten));
+            }
+        }
+
+        /// <summary>
+        /// Returns a <see cref="Memory{T}"/> to write to that is at least the requested size, as specified by the <paramref name="sizeHint"/> parameter.
+        /// </summary>
+        /// <param name="sizeHint">The minimum length of the returned <see cref="Memory{T}"/>. If less than 256, a buffer of size 256 will be returned.</param>
+        /// <returns>A memory buffer of at least <paramref name="sizeHint"/> bytes. If <paramref name="sizeHint"/> is less than 256, a buffer of size 256 will be returned.</returns>
+        public Memory<byte> GetMemory(int sizeHint = 0)
+        {
+            if (sizeHint < 256)
+                sizeHint = 256;
+
+            int sizeToRent = sizeHint > _bufferSize ? sizeHint : _bufferSize;
+
+            if (_buffers.Length == 0)
+            {
+                _buffers = new Buffer[1];
+                _buffers[0].Array = ArrayPool<byte>.Shared.Rent(sizeToRent);
+                _count = 1;
+            }
+
+            ref Buffer last = ref _buffers[_count - 1];
+            var free = last.Array.AsMemory(last.Written);
+            if (free.Length >= sizeHint)
```

Comment:
`free.Length >= sizeToRent`?

Reply:
sizeHint is the amount of data they need to write, sizeToRent is what the next buffer size would be if we don't have enough room.